### PR TITLE
Expiration perf

### PR DESCRIPTION
**Pull request by richafrank**
Look up expired futures from in-memory Futures instead of queries to the db.

### CODE REVIEW

**Comment by richafrank** [comment]
cc @jfkirk @dmichalowicz

File: zipline/utils/data.py [hunk 1]
```diff
@@ -391,3 +395,85 @@ def _update_buffer(self, frame):
...
+    """A mapping of key-value pairs sorted by key according to the sort_key
+    function provided to the mapping.  Ties from the sort_key are broken by
+    comparing the original keys.  `iter` traverses the keys in sort order.
+
+    Parameters
+    ----------
+    mapping : mapping, optional
+    **kwargs
+        The initial mapping. This will double as the starting mapping so
+        the values must also be hashable.
```
Comment (llllllllll) [inline zipline/utils/data.py]: The values do not need to be hashable here.

Comment (llllllllll) [inline zipline/utils/data.py]: also you should add the `key` arg.

File: zipline/utils/data.py [hunk 2]
```diff
@@ -391,3 +395,85 @@ def _update_buffer(self, frame):
...
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(-1, 'negative one'),
+                                         (1, 'one'),
+                                         (2, 'two')])
+    >>> del d[2]
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(-1, 'negative one'),
+                                         (1, 'one')])
+    """
+    def __init__(self, sort_key, mapping=None, **kwargs):
```
Comment (llllllllll) [inline zipline/utils/data.py]: I would just call this `key` to match `sorted`

File: zipline/utils/data.py [hunk 3]
```diff
@@ -391,3 +395,85 @@ def _update_buffer(self, frame):
...
+
+    def __len__(self):
+        return len(self._map)
+
+    def __repr__(self, _repr_running={}):
+        # Based on OrderedDict/defaultdict
+        call_key = id(self), get_ident()
+        if call_key in _repr_running:
+            return '...'
+        _repr_running[call_key] = 1
```
Comment (llllllllll) [inline zipline/utils/data.py]: too bad there isn't reprlib in py2

File: zipline/utils/data.py [hunk 4]
```diff
@@ -391,3 +395,85 @@ def _update_buffer(self, frame):
...
+        try:
+            if not self:
+                return '%s[%r]()' % (self.__class__.__name__, self._sort_key)
+            return '%s[%r](%r)' % (self.__class__.__name__, self._sort_key,
```
Comment (llllllllll) [inline zipline/utils/data.py]: I would make the format `%s(%r, %r)` so you could reevaluate it.

File: zipline/utils/data.py (hunk 1 above)
Comment (richafrank) [inline zipline/utils/data.py]: Agreed, had mis-copied.

File: zipline/utils/data.py (hunk 2 above)
Comment (richafrank) [inline zipline/utils/data.py]: Sounds good to me.

Comment (llllllllll) [inline zipline/utils/data.py]: do you think it makes sense to default `key` to `operator.le`?

File: zipline/utils/data.py (hunk 4 above)
Comment (richafrank) [inline zipline/utils/data.py]: Good call.

File: zipline/utils/data.py [hunk 5]
```diff
@@ -391,3 +399,84 @@ def _update_buffer(self, frame):
...
+    def __iter__(self):
+        for key, name in self._sorted_key_names:
+            yield name
+
+    def __len__(self):
+        return len(self._map)
+
+    def __repr__(self, _repr_running={}):
+        # Based on OrderedDict/defaultdict
+        call_key = id(self), get_ident()
```
Comment (llllllllll) [inline zipline/utils/data.py]: `_get_ident`

File: zipline/utils/data.py (hunk 4 above)
Comment (llllllllll) [inline zipline/utils/data.py]: also in py3 this only shows: `SortedDict[<function <lambda> at 0x7fae9a31cf28>](ItemsView(...))`, we might want `list(self.items())`

File: zipline/utils/data.py [hunk 6]
```diff
@@ -391,3 +399,84 @@ def _update_buffer(self, frame):
...
+        self._map[name] = value
+        if len(self._map) > len(self._sorted_key_names):
+            key = self._sort_key(name)
+            pair = (key, name)
+            idx = _bisect_right(self._sorted_key_names, pair)
+            self._sorted_key_names.insert(idx, pair)
+
+    def __delitem__(self, name, _bisect_right=bisect.bisect_right):
+        del self._map[name]
+        idx = _bisect_right(self._sorted_key_names,
```
Comment (llllllllll) [inline zipline/utils/data.py]: is this just a `bisect_left`?

File: zipline/utils/data.py [hunk 7]
```diff
@@ -391,3 +399,84 @@ def _update_buffer(self, frame):
...
+        del self._sorted_key_names[idx - 1]
+
+    def __iter__(self):
+        for key, name in self._sorted_key_names:
+            yield name
+
+    def __len__(self):
+        return len(self._map)
+
+    def __repr__(self, _repr_running={}):
```
Comment (llllllllll) [inline zipline/utils/data.py]: I think we are just using `_repr_running` as a `set`. We only check the keys and put in a single 1 constant. I think the intent might be more clear that way.

File: zipline/utils/data.py (hunk 4 above)
Comment (richafrank) [inline zipline/utils/data.py]: Agreed.

File: zipline/utils/data.py (hunk 2 above)
Comment (richafrank) [inline zipline/utils/data.py]: No strong opinion...

**Pull request merged**